README.rst
Nagios sql
----------

Install:
========

Requires python3, if you are on centos/redhat you can use: https://www.softwarecollections.org/en/ and install python3.

From pip::

    pip3 install nagios_sql

From pip with github::

    pip3 install git+https://github.com/pablodav/nagios_sql.git

For development::

     # clone the repository
     python3 setup.py develop

Usage:
======

Get help and options::

    nagios_sql --help

    usage: nagios_sql [-h] [-H HOST]
                      [--test {db_state,logship_status,mirror_status,replication_status,sql_ping,synchronization_databases_ag}]
                      [-U USER] [-P PASSWORD] [-v [VERSION]]

    optional arguments:
      -h, --help            show this help message and exit
      -H HOST, --host HOST  SQL Backup HOST to connect
      --test {db_state,logship_status,mirror_status,replication_status,sql_ping,synchronization_databases_ag}, -t {db_state,logship_status,mirror_status,replication_status,sql_ping,synchronization_databases_ag}
                            tests:
                              db_state
                              logship_status
                              mirror_status
                              replication_status
                              sql_ping
                              synchronization_databases_ag
      -U USER, --user USER  User to auth to DB
      -P PASSWORD, --password PASSWORD
                            Password to auth to DB
      -v [VERSION], --version [VERSION]
                            Gets version number


Check replication_status::

    nagios_sql -H SERVERNAME -U 'USERNAME' -P 'PASSWORD' -t replication_status -N 'publisher_name'

    CRITICAL: Replication CRITICAL
    OK Pub:Test_Replication1 DB:Test_DB1 Status:Idle MaxLatency:31s
    CRITICAL Pub:Test_Replication1_2 DB:Test_DB1 Status:Failed MaxLatency:31s
    CRITICAL Pub:Test_Replication1_3 DB:Test_DB1 Status:Failed MaxLatency:31s
    OK Pub:Test_Replication1_4 DB:Test_DB1 Status:Idle MaxLatency:31s
    OK Pub:Test_Replication1_5 DB:Test_DB1 Status:Idle MaxLatency:31s
    CRITICAL Sub:SERVERNAME DB:Test_DB1_Reporting Status:Failed Latency:?s
    CRITICAL Sub:SERVERNAME DB:Test_DB1_Reporting Status:Failed Latency:?s
    OK Sub:SERVERNAME DB:Test_DB1_Reporting Status:Idle Latency:0s
    OK Sub:SERVERNAME DB:Test_DB1_Reporting Status:Idle Latency:0s
    OK Sub:SERVERNAME DB:Test_DB1_Reporting Status:Idle Latency:0s
	
Check availability_group_status:

    nagios_sql -H SERVERNAME -U 'USERNAME' -P 'PASSWORD' -t synchronization_databases_ag -N 'publisher_name'
	
	OK: Availability Group OK
	Group:Server_AG Primary Replica:Server01 State:HEALTHY
	
Check synchronization_databases_ag:

    nagios_sql -H SERVERNAME -U 'USERNAME' -P 'PASSWORD' -t synchronization_databases_ag -S 'server_primary'
	
	Ok: Name:Replication01 State:SYNCHRONIZING Health:HEALTHY
	Name:Name:Replication02 State:SYNCHRONIZING Health:HEALTHY
	Name:Name:Replication03 State:SYNCHRONIZING Health:HEALTHY

Setup nagios command:
=====================

Configure you own `commands_nagios.cfg`::

    # standard way to define a command
    define command{
        command_name	check_nagios_sql
        command_line	nagios_sql -H $HOSTADDRESS$ -U $ARG1$ -P $ARG2$ -t $ARG3$
    }

    # this other custom settings will allow you to parse any argument
    # like:
    # command  check_nagios_sql_custom!"-U 'USERNAME' -P 'PASSWORD' -t replication_status"
    define command{
        command_name	check_nagios_sql_custom
        command_line	nagios_sql -H $HOSTADDRESS$ $ARG1$
    }

Automatic provisioned with ansible by: https://github.com/CoffeeITWorks/ansible_nagios4_server_plugins

Setup nagios service:
=====================

Example configure your service `sql_replicas.cfg`::

    # Nagios checks for SQL replicas
    # group defined on groups.cfg
    # https://github.com/pablodav/nagios_sql

    define service {
        hostgroup_name                  sql_servers_replicas
        service_description             sql_replicas_status
        check_command                   check_nagios_sql!'username'!'password'!replication_status
        notes                           some useful notes for your monitoring team
        use                             generic-service
    }


Original author:
================

original code: http://code.activestate.com/recipes/577599-nagios-plugin-for-monitoring-database-servers/
Nagios_sql.py - Matt Keranen 2011 (mksql@yahoo.com)

Author:
=======

This script was refactored and also a python package was created by:

Pablo Estigarribia 201705 (pablodav at gmail)

Collaborators:
==============

--- put your name here ---

Troubleshooting replicas
========================

Some time when you are monitoring replicas, you can see some publication that doesn't exist anymore but they still
appears in `distribution` database.

The unique way to fix them is to drop the publication, but as it doesn't exist: SQL will fail trying to drop.
So here there are some steps to create and drop the publication:

This example is only for databases that already have some other publications working, but you need to create and drop
an missing publication that still appears in `distribution` database.

.. code-block:: sql

    -- Adding the transactional publication
    use [databasename]
    exec sp_addpublication @publication = N'MyReplPub',
    @description = N'Transactional publication of database ''databasename'' from Publisher ''servername''.',
    @sync_method = N'concurrent', @retention = 0, @allow_push = N'true', @allow_pull = N'true', @allow_anonymous = N'true',
    @enabled_for_internet = N'false', @snapshot_in_defaultfolder = N'true', @compress_snapshot = N'false', @ftp_port = 21,
    @ftp_login = N'anonymous', @allow_subscription_copy = N'false', @add_to_active_directory = N'false', @repl_freq = N'continuous',
    @status = N'active', @independent_agent = N'true', @immediate_sync = N'true', @allow_sync_tran = N'false',
    @autogen_sync_procs = N'false', @allow_queued_tran = N'false', @allow_dts = N'false', @replicate_ddl = 1,
    @allow_initialize_from_backup = N'false', @enabled_for_p2p = N'false', @enabled_for_het_sub = N'false'
    GO


    exec sp_droppublication @publication = N'MyReplPub'

In some strange case you could have orphaned publication in a database that has no publications, sql will give you error saying the database is not enabled for publications when running the above command. You could try to enable, run the above command and then disable. Use this command to enable with `true` and disable with `false`

.. code-block:: sql

    use master
    exec sp_replicationdboption @dbname = N'databasename', @optname = N'publish', @value = N'true'
    GO


Nice references:
================

https://www.mssqltips.com/sqlservertip/2710/steps-to-clean-up-orphaned-replication-settings-in-sql-server/

Other ways of deployment:
=========================

This plugin is already included in ansible role: https://github.com/CoffeeITWorks/ansible_nagios4_server_plugins 



CHANGELOG
0.9.0
    * Hotfix - Check publications and subscriptions
0.8.1
    * workaround to allow dependency with pymssql
0.8.0
    * Added checks for synchronization database ag on clusters
0.6.0
    * Added checks for availability group
    * fixes for replication_status with different publisher
0.5.1
    * fixed homepage
0.5.0
    * first version with refactor from original script.


src/nagios_sql.py
#!/usr/bin/env python

# original code: http://code.activestate.com/recipes/577599-nagios-plugin-for-monitoring-database-servers/
# Nagios_sql.py - Matt Keranen 2011 (mksql@yahoo.com)
# Pablo Estigarribia 201705 (pablodav at gmail)
#
# changes:
# * use pymssql instead of pyodbc to have simple setup
# * refactor compatible to python3
# * code inspect and fixes to pep8 compliant
# * Added possibility to use user and password from cmd
# * Added better output on CRITICAL and WARNING in replication_status

import sys
import pymssql
import argparse
from . import __version__


nagios_codes = dict(OK=0, WARNING=1, CRITICAL=2, UNKNOWN=3, DEPENDENT=4)


def parse_args(args):
    """
    Information extracted from: https://mkaz.com/2014/07/26/python-argparse-cookbook/
    :return:
    """
    parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)

    parser.add_argument('-H', '--host', dest='host',
                        help='SQL Backup HOST to connect')

    # Adding test choices
    parser.add_argument("--test", '-t', dest='test',
                        choices=test_pylist(),
                        help=test_list()
                        )

    parser.add_argument("-U", "--user", dest='user', help="User to auth to DB")
    parser.add_argument("-P", "--password", dest='password', help="Password to auth to DB")
    parser.add_argument("-S", "--server_primary", dest='server_primary', help="Specify primary availability group server")
    parser.add_argument("-N", "--publisher_name", dest='publisher_name', 
                        help='used to monitor replication_status in diffent publisher',
                        default=None)
	
    parser.add_argument("-v", "--version", dest='version', nargs='?', default=None, const=True,
                        help="Gets version number")

    if not args:
        raise SystemExit(parser.print_help())

    return parser.parse_args(args)


def nagios_return(code, response):
    print((code + ": " + response))
    sys.exit(nagios_codes[code])


def execute_sql(host, sql, database='master', user="user", password="password", return_dict=True):
    """Execute SQL against specified database"""

    user = user
    password = password
    server = host

    try:
        conn = pymssql.connect(server=server,
                               user=user,
                               password=password,
                               database=database)
    except pymssql.Error as e:
        return {'code': 'CRITICAL', 'msg': "Unable to connect to SQL host %s\n%s" % (host, e)}

    if return_dict:
        cursor = conn.cursor(as_dict=True)
    else:
        cursor = conn.cursor()

    cursor.execute(sql)
    rows = cursor.fetchall()

    cursor.close()
    conn.close()

    return rows


def get_func(test):
    """Determine if function name is valid and defined as a Nagios test"""

    try:
        func = getattr(sys.modules[__name__], test)
    except AttributeError as e:
        nagios_return('UNKNOWN', 'Invalid test name %s' % test)

    try:
        test = func.is_test
    except:
        nagios_return('UNKNOWN', '%s not defined as test' % test)

    if test:
        return func
    else:
        nagios_return('UNKNOWN', '%s not defined as test' % test)


def test_pylist():
    """Python List of valid test names"""
    tests = []
    for func in dir(sys.modules[__name__]):
        test = getattr(sys.modules[__name__], func)
        try:
            test.is_test
        except:
            pass
        else:
            tests.append(func)

    return tests


def test_list():
    """List of valid test names"""
    tests = 'tests:\n'
    for func in dir(sys.modules[__name__]):
        test = getattr(sys.modules[__name__], func)
        try:
            test.is_test
        except:
            pass
        else:
            tests += '  %s\n' % func

    return tests


def nagios_test(func):
    func.is_test = True
    return func


@nagios_test
def sql_ping(host, user, password):
    """Connect to SQL Server instance and return version string"""

    sql = 'SELECT @@VERSION'
    rows = execute_sql(host, sql, user=user, password=password, return_dict=False)
    if type(rows) is dict:
        return rows

    message = rows[0][0]

    return {'code': 'OK', 'msg': message}


@nagios_test
def db_state(host, user, password):
    """Check state of each database"""
    crit = warn = 0
    msg = ''

    sql = 'SELECT [name] db, user_access, user_access_desc, state, state_desc FROM sys.databases ' \
          'WHERE user_access > 0 OR state > 0'
    rows = execute_sql(host, sql, user=user, password=password)
    if type(rows) is dict:
        return rows

    for row in rows:
        if row.get("state") == 6:
            warn += 1
            msg += 'Database %s is %s\n' % (row.get("db"), row.get("state_desc"))
        elif row.get("state") > 3:
            crit += 1
            msg += 'Database %s is %s\n' % (row.get("d"), row.get("state_desc"))

        if row.get("user_access") > 0:
            # warn += 1
            msg += 'Database %s in mode %s\n' % (row.get("db"), row.get("user_access_desc"))

    if crit > 0:
        code = 'CRITICAL'
        msg = 'Database state CRITICAL\n' + msg
    elif warn > 0:
        code = 'WARNING'
        msg = 'Database state warning\n' + msg
    else:
        code = 'OK'
        msg = 'Databases OK\n' + msg

    return {'code': code, 'msg': msg}


@nagios_test
def replication_status(host, user, password, publisher_name='@@SERVERNAME'):
    """Report transactional replication status"""
    # 0 = error ('Unknown')
    # mstat: status of check publication or subscription
    # mwarn: warning status received from check of publication or subscription
    # Default 0 for both
    mstat = mwarn = 0
    msg = ''
    status_critical_list = [5, 6]
    status_ok_list = [1, 2, 3, 4]

    # Status dictionaries 
    status = {0: 'Unknown', 1: 'Started', 2: 'Succeeded', 3: 'Active', 4: 'Idle', 5: 'Retrying', 6: 'Failed'}
    warning = {0: '', 1: '-Expiration ', 2: '-Latency '}

    def __get_publications_status():
        """
        Check if publication status is not critical and returns status number.
        Check if publication warning is not critical and returns mwarn number.
        Adds msg and publication status vars.

        return: mstat, mwarn, msg
        """
        _mstat = _mwarn = 0
        _msg = ''
        

        # Transactional replication - publications
        # Get data
        sql = 'EXEC dbo.sp_replmonitorhelppublication @publisher = {}'.format(publisher_name)
        rows = execute_sql(host, sql, 'distribution', user=user, password=password)
        if type(rows) is dict:
            return rows

        # Check publications
        for row in rows:

            publication_status = 'OK'

            if row.get("status") in status_critical_list:
                publication_status = "CRITICAL"
                _mstat = row["status"]
                
            elif row.get("warning") > 0:
                publication_status = "WARNING"
                _mwarn = row["warning"]

            if row.get("worst_latency") is None:
                row["worst_latency"] = 0

            _msg += '%s Pub:%s DB:%s Status:%s%s MaxLatency:%ss\n' % (
                publication_status,
                row["publication"],
                row["publisher_db"],
                status.get(row["status"]),
                warning.get(row["warning"]),
                row["worst_latency"])
            
        return _mstat, _mwarn, _msg

    def __get_subscription_status():
        """
        return: mstat, mwarn, msg
        """
        _mstat = _mwarn = 0
        _msg = ''
        # Get data
        sql = 'EXEC dbo.sp_replmonitorhelpsubscription @publisher = {}'.format(publisher_name) + ', @publication_type = 0'
        rows = execute_sql(host, sql, 'distribution', user=user, password=password)
        if type(rows) is dict:
            return rows

        # Check suscriptions
        # will set _mstat or _mwarn to CRITICAL or WARNING number value if any error found
        for row in rows:

            subscription_status = "OK"

            if row.get("status") in status_critical_list:
                subscription_status = "CRITICAL"
                _mstat = row["status"]
            elif row.get("warning") > 0:
                subscription_status = "WARNING"
                _mwarn = row["warning"]

            if row.get("latency") is None:
                row["latency"] = '?'

            _msg += '%s Sub:%s DB:%s Status:%s%s Latency:%ss\n' % (
                subscription_status,
                row.get("subscriber"),
                row.get("subscriber_db"),
                status.get(row["status"]),
                warning.get(row["warning"]),
                row.get("latency"))
        
        return _mstat, _mwarn, _msg

    # Check publications status
    # change variables received from the check.
    mstat, mwarn, msg = __get_publications_status()

    # Transactional replication - suscriptions
    # Do not check subscription_status if publication is CRITICAL.    
    if mstat not in status_critical_list:
        mstat, mwarn, _msg_subscription = __get_subscription_status()
        msg += _msg_subscription

    if mstat in status_critical_list:
        code = 'CRITICAL'
        msg = 'Replication CRITICAL\n' + msg
    elif mwarn > 0:
        code = 'WARNING'
        msg = 'Replication WARNING\n' + msg
    elif mstat in status_ok_list:
        code = 'OK'
        msg = 'Replication OK\n' + msg
    else:
        code = 'CRITICAL'
        msg = 'Replication CRITICAL\n' + msg

    return {'code': code, 'msg': msg}


@nagios_test
def mirror_status(host, user, password):
    """Report mirror status"""
    crit = warn = 0
    msg = ''

    sql = """SELECT d.name dbname, m.mirroring_partner_instance partner, m.mirroring_state, m.mirroring_state_desc state
        FROM sys.databases d
        INNER JOIN sys.database_mirroring m ON m.database_id = d.database_id
        WHERE m.mirroring_state IS NOT NULL"""
    rows = execute_sql(host, sql, user=user, password=password)
    if type(rows) is dict:
        return rows

    # state = {0:'Suspended', 1:'Disconnected', 2:'Synchronizing', 3:'PendingFailover', 4:'Synchronized'}
    # sql = "EXEC sp_dbmmonitorresults '%s'" % dbname

    for row in rows:
        if row.get("mirroring_state") < 2:
            crit += 1
        if row.get("mirroring_state") == 3:
            warn += 1

        msg += "DB:%s Partner:%s State:%s\n" % (
            row.get("dbname"),
            row.get("partner"),
            row.get("state"))

    if crit > 0:
        code = 'CRITICAL'
        msg = 'Mirroring CRITICAL\n' + msg
    elif warn > 0:
        code = 'WARNING'
        msg = 'Mirroring warning\n' + msg
    else:
        code = 'OK'
        msg = 'Mirroring OK\n' + msg

    return {'code': code, 'msg': msg}


@nagios_test
def logship_status(host, user, password):
    """Report log shipping retstore delta and latency"""
    crit = warn = 0
    msg = ''

    sql = """SELECT secondary_server, secondary_database, primary_server, primary_database,
        last_restored_date, DATEDIFF(mi, last_restored_date, GETDATE()) last_restored_delta,
        last_restored_latency, restore_threshold
        FROM msdb..log_shipping_monitor_secondary"""
    rows = execute_sql(host, sql, user=user, password=password)
    if type(rows) is dict:
        return rows

    for row in rows:
        if row.get("last_restored_delta") >= row.get("restore_threshold"):
            warn += 1
            msg += "Srv:%s DB:%s Restore delta %s exceeds threshold of %s\n" % (
                row.get("primary_server"),
                row.get("primary_database"),
                row.get("last_restored_delta"),
                row.get("restore_threshold"))
        if row.get("last_restored_latency") >= row.get("restore_threshold"):
            crit += 1
            msg += "Srv:%s DB:%s Latency of %s exceeds threshold of %s\n" % (
                row.get("primary_server"),
                row.get("primary_database"),
                row.get("last_restored_latency"),
                row.get("restore_threshold"))
        if row.get("last_restored_delta") < row.get("restore_threshold") and \
           row.get("last_restored_latency") < row.get("restore_threshold"):
            msg += "Srv:%s DB:%s Latency:%s Restore delta:%s\n" % (
                row.get("primary_server"),
                row.get("primary_database"),
                row.get("last_restored_latency"),
                row.get("last_restored_delta"))

    if crit > 0:
        code = 'CRITICAL'
        msg = 'Log shipping CRITICAL\n' + msg
    elif warn > 0:
        code = 'WARNING'
        msg = 'Log shipping warning\n' + msg
    else:
        code = 'OK'
        msg = 'Log shipping OK\n' + msg

    return {'code': code, 'msg': msg}
    
@nagios_test
def availability_group_status(host, user, password):
    """availability group status"""
    crit = warn = 0
    msg = ''

    sql = """SELECT ag.name agname, ags.* FROM sys.dm_hadr_availability_group_states ags INNER JOIN sys.availability_groups ag ON ag.group_id = ags.group_id"""
    rows = execute_sql(host, sql, user=user, password=password)
    if type(rows) is dict:
        return rows
    
    #https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/sys-dm-hadr-availability-replica-states-transact-sql?view=sql-server-2017
    # state = {0:'Not healthy', 1:'Partially healthy', 2:'Healthy'}

    for row in rows:
        if row.get("synchronization_health") == 0:
            crit += 1
        if row.get("synchronization_health") == 1:
            warn += 1

        msg += "Group:{} Primary Replica:{} State:{}\n".format(
            row.get("agname"),
            row.get("primary_replica"),
            row.get("synchronization_health_desc"))

    if crit > 0:
        code = 'CRITICAL'
        msg = 'Availability Group CRITICAL\n' + msg
    elif warn > 0:
        code = 'WARNING'
        msg = 'Availability Group warning\n' + msg
    else:
        code = 'OK'
        msg = 'Availability Group OK\n' + msg

    return {'code': code, 'msg': msg}

@nagios_test
def synchronization_databases_ag(host, user, password, server_primary):
    """Databases Availability group Status"""
	
    msg = ''
    primary = False
    synchronized = 'OK'
	
    #status = {0: 'Unknown', 1: 'Started', 2: 'Succeeded', 3: 'Active', 4: 'Idle', 5: 'Retrying', 6: 'Failed'}
    #warning = {0: '', 1: '-Expiration ', 2: '-Latency '}

    sql = """select * FROM sys.dm_hadr_availability_group_states"""
    rows = execute_sql(host, sql, user=user, password=password)
	
    for row in rows:
        if row.get("primary_replica") == server_primary:
            primary = True

    sql = """SELECT db.name,repstat.synchronization_state_desc,repstat.synchronization_health_desc FROM sys.databases db INNER JOIN sys.dm_hadr_database_replica_states repstat ON db.database_id = repstat.database_id WHERE is_local = 1"""
    rows = execute_sql(host, sql, user=user, password=password)

    if type(rows) is dict:
        return rows
	
    if primary:

        for row in rows:
        
            if row.get("synchronization_state_desc") != "SYNCHRONIZED":
                synchronized = "CRITICAL"
            
            msg += "Name:{} State:{} Health:{}\n".format(
                row.get("name"),
                row.get("synchronization_state_desc"),
                row.get("synchronization_health_desc"))

    else:
        
        for row in rows:
            
            if row.get("synchronization_state_desc") != "SYNCHRONIZING":
                synchronized = "CRITICAL"
            
            msg += "Name:{} State:{} Health:{}\n".format(
                row.get("name"),
                row.get("synchronization_state_desc"),
                row.get("synchronization_health_desc"))
    
    msg = 'Database synchronization\n' + msg
    	
    return {'code': synchronized, 'msg': msg}


def main():

    options = parse_args(sys.argv[1:])

    if options.version:
        raise SystemExit('{}'.format(__version__))

    host = options.host
    test = options.test
    user = options.user
    password = options.password
    publisher_name = options.publisher_name
    server_primary = options.server_primary

    func = get_func(test)
    if publisher_name:
        result = func(host, user=user, password=password, publisher_name=publisher_name)
    elif server_primary:
        result = func(host, user=user, password=password, server_primary=server_primary)
    else:
        result = func(host, user=user, password=password)

    nagios_return(result['code'], result['msg'])
    

if __name__ == "__main__":
    main()
    
